mqtt/mqtt-bridge/src/client.rs

```diff
@@ -0,0 +1,227 @@
+            ),
+            Credentials::Anonymous(client_id) => (client_id.into(), None),
+        };
+
+        let client = if clean_session {
+            Client::new(
+                Some(client_id),
+                None,
+                None,
+                TcpConnection {
+                    address: address.into(),
+                    token_source,
+                },
+                DEFAULT_MAX_RECONNECT,
+                keep_alive,
+            )
+        } else {
+            Client::from_state(
+                client_id,
+                None,
+                None,
+                TcpConnection {
+                    address: address.into(),
+                    token_source,
+                },
+                DEFAULT_MAX_RECONNECT,
+                keep_alive,
+            )
+        };
+
+        MqttClient {
+            client,
+            event_handler,
+        }
+    }
+
+    pub fn shutdown_handle(&self) -> Result<ShutdownHandle, ShutdownError> {
+        self.client
+            .shutdown_handle()
+            .map_or(Err(ShutdownError::ClientDoesNotExist), |shutdown_handle| {
+                Ok(ShutdownHandle(shutdown_handle))
+            })
+    }
+
+    pub async fn handle_events(mut self) -> Result<(), ClientConnectError> {
+        while let Some(event) = self.client.try_next().await.unwrap_or_else(|e| {
+            error!(message = "failed to poll events", error=%e);
+            None
+        }) {
+            if let Err(e) = self.event_handler.handle_event(event).await {
+                error!("error processing event {}", e);
+            }
+        }
+
+        Ok(())
+    }
+
+    pub async fn subscribe(&mut self, topics: Vec<String>) -> Result<(), ClientConnectError> {
+        debug!("subscribing to topics");
+        let subscriptions = topics.iter().map(|topic| proto::SubscribeTo {
+            topic_filter: topic.to_string(),
+            qos: proto::QoS::AtLeastOnce,
+        });
+
+        for subscription in subscriptions {
+            debug!("subscribing to topic {}", subscription.topic_filter);
+            self.client
+                .subscribe(subscription)
+                .map_err(ClientConnectError::SubscribeFailure)?;
+        }
+
+        let mut subacks: HashSet<_> = topics.iter().collect();
+        if subacks.is_empty() {
+            debug!("no topics to subscribe to");
+            return Ok(());
+        }
+
+        while let Some(event) = self
+            .client
+            .try_next()
+            .await
+            .map_err(ClientConnectError::PollClientFailure)?
+        {
+            if let Event::SubscriptionUpdates(subscriptions) = event {
+                for subscription in subscriptions {
+                    if let SubscriptionUpdateEvent::Subscribe(sub) = subscription {
+                        subacks.remove(&sub.topic_filter);
+                    }
+                }
+
+                if subacks.is_empty() {
```

Comment:
It should be OK to receive NotAllowed from a broker, right? I think we need to log error and continue, instead of erroring out client

Reply:
the way the mqtt client is implemented it would not send the error to the caller so actually it would keep looping if the sub is not allowed. So a put a temporary fix, but will need to modify the client to send an event if the sub is not allowed.

Reply:
Yes, but also in addition to what you said, it is not an error, when the bridging client cannot subscribe to a broker, right? lets say it wants to sub to both "foo" and "bar" topics. Even if the broker allows only "foo" and not "bar" the bridge should not stop with an error. I should log an error and continue to work.

bridge behavior in this case is different from what CommandHandler has. CommandHandler errors out when it cannot subscribe to **all** required topics.